src/net_processing.cpp

```diff
@@ -2225,6 +2245,9 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
     {
         std::vector<CBlockHeader> headers;
 
+        if (pfrom->nPendingHeaderRequests > 0)
+          pfrom->nPendingHeaderRequests -= 1;
```

Comment:
Just a general comment about `nPendingHeaderRequests` in this pull request. While the reads and writes to `nPendingHeaderRequests` are thread safe due to it being atomic, the logic itself is not necessarily thread safe. This might not even be an issue here due to the majority of the logic already executing with a locked cs_main

```
        if (pfrom->nPendingHeaderRequests > 0)
          pfrom->nPendingHeaderRequests -= 1;
```

Can multiple threads execute this at the same time? If so, then multiple threads could all read the value of `nPendingHeaderRequests` at the same time, seeing a value of 1 for instance, then the threads can all decrement that value, resulting in a negative `nPendingHeaderRequests`. 

It looks like `cs_main` is always locked when calling `RequestHeadersFrom` so I don't see anything really popping up in there with the same sort of issue.

Reply:
Thank you. `ProcessMessage()` is currently executed in a single `msghandler` thread, but I think you're right regardless of that current implementation and it would be a good idea to perform a lock here too. 

Will address. Thanks!

Reply:
I have created issue #2425 because I have more than just this code to look into re: thread safety / future proofing.